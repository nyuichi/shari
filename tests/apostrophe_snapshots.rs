use std::fmt::Write as _;

fn format_error(err: &anyhow::Error) -> String {
    let mut lines = vec!["error chain:".to_string()];
    for cause in err.chain() {
        lines.push(format!("  - {cause}"));
    }
    lines.join("\n")
}

#[test]
fn apostrophe_identifiers_snapshot() {
    let cases = [
        (
            "const_with_apostrophe",
            "type const Prop : Type\nconst foo' : Prop\n",
        ),
        (
            "axiom_with_apostrophe",
            "type const Prop : Type\nconst foo : Prop\naxiom bar' : foo\n",
        ),
        (
            "type_const_with_apostrophe",
            "type const Prop : Type\ntype const P' : Type\n",
        ),
        (
            "const_starting_with_apostrophe",
            "type const Prop : Type\nconst 'x : Prop\n",
        ),
    ];

    let mut report = String::new();

    for (name, input) in cases {
        writeln!(report, "{name}:").expect("write succeeds");
        match shari::process(input) {
            Ok(()) => {
                writeln!(report, "  ok").expect("write succeeds");
            }
            Err(err) => {
                writeln!(report, "  error").expect("write succeeds");
                for line in format_error(&err).lines() {
                    writeln!(report, "    {line}").expect("write succeeds");
                }
            }
        }
        writeln!(report).expect("write succeeds");
    }

    insta::assert_snapshot!("apostrophe_identifiers", report);
}
