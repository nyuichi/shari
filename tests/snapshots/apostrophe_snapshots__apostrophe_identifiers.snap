---
source: tests/apostrophe_snapshots.rs
assertion_line: 50
expression: report
---
const_with_apostrophe:
  ok

axiom_with_apostrophe:
  ok

type_const_with_apostrophe:
  ok

const_starting_with_apostrophe:
  error
    error chain:
      - parse error
      - parse error: expected identifier at 2:7
    
    const 'x : Prop
          ^
